### ISSUE TITLE
How to Avoid local minimum

### ISSUE THREAD

**Issue by mwindowshz** [AlexeyAB/darknet#307]
Hi
I am training a network, using VOC image and my own images.
I run it for 50,000 iterations.
with yolo-voc.2.0.cfg without random.

It seems like the learning is stuck and cannot get to a small minimum.
the avg values of go down until they are stuck around 0.2-0.16 value.
![image](https://user-images.githubusercontent.com/17617515/33939378-9ecce2d4-e013-11e7-9687-0509cee4b379.png)

avg on 33000 is 0.169421
avg on 50000 is 0.205171

calling recall on 33000 
 5030  9488 12385       RPs/Img: 2.83   IOU: 64.27%     Recall:76.61%
calling recall on 50000
 5030  9513 12385       RPs/Img: 2.83   IOU: 64.42%     Recall:76.81%

this is about the same.
meaning the network did not learn more.

What should I do to help the network learn more?
change learning reate? decay? momentum?

Is this value of 0.2 good?
do I need a lower value?

![image](https://user-images.githubusercontent.com/17617515/33939452-ec321508-e013-11e7-9434-36e91518bf5b.png)

**Comment by AlexeyAB** [comment]
Hi,

* Seems that your network (cfg) can't learn more on your dataset, so you can weights file after 33000 iterations, so avg_loss=0.169421 is enough in your case
* Do you use diffrent datasets for recall and for training?
* No, you should change learning_rate, decay or momentum
* You can try to train with `random=1`: https://github.com/AlexeyAB/darknet#how-to-improve-object-detection
* What average image size do you use in your dataset? And what range of sizes of your objects (in pixels)?
* You can try to train with random=0, width=608 height=608 or width=832 height=832
* You can try to base your cfg-file on the `densenet201-yolo2.cfg` or `resnet50-yolo.cfg` instead of yolo-voc.2.0.cfg: https://github.com/AlexeyAB/darknet/issues/179#issuecomment-330047738

**Comment by mwindowshz** [comment]
Hi
Thank you.
> No, you should change learning_rate, decay or momentum

Ok - I would not change them, I tried, training again from 33000, with diffrent decay momentum and learning rate, but it still did not help

I used diffrent datasets for recall and training, there is a train.txt and valid.txt with diffrent lists of images.

> What average image size do you use in your dataset? And what range of sizes of your objects (in pixels)?

My data set is VOC 2007 2012  (there average is 300x500) + My Images - 
Most of my Images are sized about 400X400 (and a few are larger 500x500 700x800)
the objects are small, for example - 40x60 , 20x40 , 50x120 30x50 pixels.  

In VOC images objects have diffrent sizes, many are large, and can be almost half or quarter of the image (300x500,150x300). and otheres are smaller 100x50 


> You can try to train with random=0, width=608 height=608 or width=832 height=832

Should I do this with my size of images? would it help the detections of small objects?

> You can try to base your cfg-file on the densenet201-yolo2.cfg or resnet50-yolo.cfg instead of yolo-voc.2.0.cfg: #179 (comment)

Ok I would try this out!

Thank you very much!

**Comment by AlexeyAB** [comment]
> > You can try to train with random=0, width=608 height=608 or width=832 height=832

> Should I do this with my size of images? would it help the detections of small objects?

1. use `random=0 width=832 height=832` for training and detection will help to **detect very small objects - it will reduce speed of training and detection**
2. use `random=1 width=416 height=416` for training and detection will help to **detect small and large objects - it will reduce speed of training only**
3. use `random=1 width=608 height=608` - slightly more accurate and slightly slower than point 2

**Comment by mwindowshz** [comment]
Hi 
I am trying your recommendations.

Should I also check into anchors?
do I need to calculate them for my Data set?
Is there a build in method or a guide for this?

Also question, if the images are small, then resizing them will cause blurry details  in the image? so can this cause problems?
Also at run time sometimes I have small images (100,100 , 80,200) that I receive so resizing could cause more blurry details. So I guess there would be a need to find balance.

> detect very small objects

What are considered "very small objects" , are the sizes I described very small? 

>  40x60 , 20x40 , 50x120 30x50 pixels

or you mean even smaller such as 10x10 pixels? 5x5?  

Thanks.

**Comment by AlexeyAB** [comment]
@mwindowshz Very small - when object size small than 16x16 after that image is resized to size of network.
